fs: fat_fs: Changing mkfs options to create FAT32 on larger storage

msnahal opened the pull request:
Changing mkfs options to create FAT32 on larger storage when FAT16 fails.

Signed-off-by: Nahal Farhi <nahal@whisper.ai>

de-nordic reviewed (changes requested):
The `FM_ANY` also adds support for EXFAT (it is `FS_FAT | FS_FAT32 | FS_EXFAT` by definition.
Selection of supported FAT for formatting should be done with added Kconfig option, that will allow to select max supported FAT system, and should also have dependency, for FS_EXFAT, from `CONFIG_FS_FATFS_EXFAT`, the option that enables EXFAT support within ELM FAT driver.

msnahal commented:
> The `FM_ANY` also adds support for EXFAT (it is `FS_FAT | FS_FAT32 | FS_EXFAT` by definition. Selection of supported FAT for formatting should be done with added Kconfig option, that will allow to select max supported FAT system, and should also have dependency, for FS_EXFAT, from `CONFIG_FS_FATFS_EXFAT`, the option that enables EXFAT support within ELM FAT driver.

[mkfs in zephyr fatfs module](https://github.com/zephyrproject-rtos/fatfs/blob/master/ff.c#L5802) does special handling of exfat when CONFIG_FS_FATFS_EXFAT is set ([here](https://github.com/zephyrproject-rtos/fatfs/blob/master/include/ffconf.h#L279), [here](https://github.com/zephyrproject-rtos/fatfs/blob/master/ff.c#L5919), [here](https://github.com/zephyrproject-rtos/fatfs/blob/master/ff.c#L5939)). So unless `CONFIG_FS_FATFS_EXFAT` is set, mkfs won't create a exfat with `FM_ANY`.

When `CONFIG_FS_FATFS_EXFAT` is not set, with option `FM_ANY`, mkfs first attempts to create FAT16 but if the storage is big, it [switches to FAT32](https://github.com/zephyrproject-rtos/fatfs/blob/master/ff.c#L6159). If the option is `FM_FAT`, mkfs only tries to create a FAT16 and it fails to do that on larger storage and [returns with an error](https://github.com/zephyrproject-rtos/fatfs/blob/master/ff.c#L6162). @de-nordic do you think we need a kconfig to select fat32 explicitly or it's fine that the driver attempts to create FAT16 and it automatically switches to FAT32 on larger storage?

de-nordic commented:
>
> When `CONFIG_FS_FATFS_EXFAT` is not set, with option `FM_ANY`, mkfs first attempts to create FAT16 but if the storage is big, it [switches to FAT32](https://github.com/zephyrproject-rtos/fatfs/blob/master/ff.c#L6159). If the option is `FM_FAT`, mkfs only tries to create a FAT16 and it fails to do that on larger storage and [returns with an error](https://github.com/zephyrproject-rtos/fatfs/blob/master/ff.c#L6162). @de-nordic do you think we need a kconfig to select fat32 explicitly or it's fine that the driver attempts to create FAT16 and it automatically switches to FAT32 on larger storage?

You are right, it is not needed.

de-nordic commented:
Fix compliance checks (https://docs.zephyrproject.org/latest/contribute/index.html)

Commit:
fs: fat_fs: Changing mkfs options to create FAT32 on larger storage

Changing mkfs options from FM_FAT to FM_ANY so that on larger storage
FAT32 instead of FAT16 is automatically created.

Signed-off-by: Nahal Farhi <nahal@whisper.ai>

de-nordic commented:
@nvlsianpu Please review.

owain-incus reviewed (commented):
Looks good to me. Before I made this change I was blocked from using EXFAT. With this change I am able to make fs on 32GB.

de-nordic commented:
@nashif Please merge. This has been accepted for long time now.
